File: Array/medium/40_combination_sum_ii/CombinationSumTwo.java
//
// Created by CaoTao, 2018/09/16
//

/*
 * LeetCode link:
 * https://leetcode.com/problems/combination-sum-ii/description/
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSumTwo {
  public List<List<Integer>> combinationSum2(int[] candidates, int target) {
    Arrays.sort(candidates);
    System.out.println(Arrays.toString(candidates));
    List<Integer> temp = new ArrayList<>();
    List<List<Integer>> result = new ArrayList<>();
    backtrack_2(result, temp, candidates, 0, target);
    return result;
  }

  public void backtrack(List<List<Integer>> result, List<Integer> temp, int[] nums, int start, int remain) {
    for (int i = start; i < nums.length; i++) {
      if (nums[i] > remain) {
        break;
      }
      if (nums[i] == remain) {
        temp.add(nums[i]);
        result.add(new ArrayList<>(temp));
        temp.remove(temp.size() - 1);
        while (i + 1 < nums.length && nums[i] == nums[i + 1]) {
          i++;
        }
        break;
      }
      temp.add(nums[i]);
      backtrack(result, temp, nums, i + 1, remain - nums[i]);
      temp.remove(temp.size() - 1);

      while (i + 1 < nums.length && nums[i] == nums[i + 1]) {
        i++;
      }
    }
  }

  public void backtrack_2(List<List<Integer>> result, List<Integer> temp, int[] nums, int start, int remain) {
    if (remain < 0) return;
    if (remain == 0) {
      result.add(new ArrayList<>(temp));
      return;
    } 
    
    for (int i = start; i < nums.length; i++) {
      temp.add(nums[i]);
      backtrack_2(result, temp, nums, i + 1, remain - nums[i]);
      temp.remove(temp.size() - 1);
      while (i + 1 < nums.length && nums[i] == nums[i + 1]) {
        i++;
      }
    }
  }

  public static void main(String[] args) {
    CombinationSumTwo obj = new CombinationSumTwo();

    int[] nums = { 10, 1, 2, 7, 6, 1, 5 };
    List<List<Integer>> result = obj.combinationSum2(nums, 8);
    System.out.println(result);
  }
}

File: Array/medium/40_combination_sum_ii/CombinationSumTwo.js
//
// Created by CaoTao, 2018/10/12
//

/*
 * LeetCode link:
 * https://leetcode.com/problems/combination-sum-ii/description/
 */

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  Array.prototype.sort.call(candidates, (a, b) => a - b)
  let tempArray = []
  let result = []

  // console.log('...candidates:', candidates)
  function helper(target, start) {
    if (target === 0) {
      result.push([...tempArray])
      return
    }

    for (let i = start; i < candidates.length; i++) {
      if (candidates[i] > target) return
      if (i != start && candidates[i] === candidates[i - 1]) {
        continue
      }
      tempArray.push(candidates[i])
      helper(target - candidates[i], i + 1)
      tempArray.pop()
    }
  }
  helper(target, 0)
  return result
}

console.log(combinationSum2([10,1,2,7,6,1,5], 8))
console.log(combinationSum2([2,5,2,1,2], 5))